assembl/graphql/section.py

```diff
@@ -0,0 +1,160 @@
+import graphene
+from graphene_sqlalchemy import SQLAlchemyObjectType
+from pyramid.httpexceptions import HTTPUnauthorized
+from pyramid.security import Everyone
+from graphene.relay import Node
+
+from assembl import models
+from assembl.auth import IF_OWNED, CrudPermissions
+from assembl.auth.util import get_permissions
+from assembl.models.section import SectionTypesEnum
+
+from .langstring import (LangStringEntry, LangStringEntryInput,
+                         langstring_from_input_entries, resolve_langstring,
+                         resolve_langstring_entries,
+                         update_langstring_from_input_entries)
+from .types import SecureObjectType
+from .utils import abort_transaction_on_exception
+
+
+SectionTypes = graphene.Enum.from_enum(SectionTypesEnum)
+
+
+class Section(SecureObjectType, SQLAlchemyObjectType):
+
+    class Meta:
+        model = models.Section
+        interfaces = (Node, )
+        only_fields = ('id', 'section_type', 'order')
+
+    title = graphene.String(lang=graphene.String())
+    title_entries = graphene.List(LangStringEntry)
+    url = graphene.String()
+
+    def resolve_title(self, args, context, info):
+        title = resolve_langstring(self.title, args.get('lang'))
+        return title
+
+    def resolve_title_entries(self, args, context, info):
+        return resolve_langstring_entries(self, 'title')
+
+
+class CreateSection(graphene.Mutation):
+    class Input:
+        title_entries = graphene.List(LangStringEntryInput, required=True)
+        section_type = graphene.Argument(SectionTypes)
+        url = graphene.String()
+        order = graphene.Float()
+
+    section = graphene.Field(lambda: Section)
+
+    @staticmethod
+    @abort_transaction_on_exception
+    def mutate(root, args, context, info):
+        cls = models.Section
+        discussion_id = context.matchdict['discussion_id']
+        user_id = context.authenticated_userid or Everyone
+        permissions = get_permissions(user_id, discussion_id)
+        allowed = cls.user_can_cls(
+            user_id, CrudPermissions.CREATE, permissions)
+        if not allowed or (allowed == IF_OWNED and user_id == Everyone):
+            raise HTTPUnauthorized()
+
+        with cls.default_db.no_autoflush as db:
+            title_entries = args.get('title_entries')
+            if len(title_entries) == 0:
+                raise Exception(
+                    'Resource titleEntries needs at least one entry')
+                # Better to have this message than
+                # 'NoneType' object has no attribute 'owner_object'
+                # when creating the saobj below if title=None
+
+            title_langstring = langstring_from_input_entries(title_entries)
+
+            kwargs = {}
+            kwargs['section_type'] = args.get('section_type', 'CUSTOM')
+            kwargs['url'] = args.get('url')
+            kwargs['order'] = args.get('order')
```

Comment:
You should set a default value for order, like `args.get('order', 10.0)`

Reply:
done